### PR TITLE
EXSWCPHIPT-139 - Added tests for hipStreamDestroy

### PR DESCRIPTION
**Pull request by FMarno**
CMakeLists.txt was updated so any changes to config_amd_*.json copied when the project was rebuilt.

Unit_hipStreamDestroy_DoubleDestroy fails, but this could just be undefined behaviour.

Unit_hipStreamDestroy_WithPendingWork was found to hang, I think this is due to hipStreamDestroy not synchronizing before destroying the stream, but I'm not sure what the intended functionality is.

Depends on #2517

### CODE REVIEW

File: tests/catch/unit/stream/hipStreamDestroy.cc [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+__global__ void waiting_kernel(int* semaphore) {
+  size_t tid = blockIdx.x * blockDim.x + threadIdx.x;
+  if (tid == 0) {
+    while (atomicCAS(semaphore, 1, 2) == 0) {
+    }
+  }
+}
+
+// NOTE possible integration test
+TEST_CASE("Unit_hipStreamDestroy_WithPendingWork") {
```
Comment (chrispaquot) [inline tests/catch/unit/stream/hipStreamDestroy.cc:74]: I don't like using timers like that as the system could be busy doing something else and all the assumed delays are off.
I don't feel like you need to add the thread to delay the completion of the kernel. You can prob do a spin loop inside the kernel or do the signaling from the host using hipHostAlloc memory instead of hipMalloc or something.

File: tests/catch/hipTestMain/config/config_amd_linux.json [hunk 2]
```diff
@@ -1,6 +1,10 @@
 {
     "DisabledTests":
 	[
+	   "# Skipped due to EXSWCPHIPT-19",
```
Comment (FMarno) [inline tests/catch/hipTestMain/config/config_amd_linux.json]: convert to HIP_SKIP_TEST

File: tests/catch/hipTestMain/config/config_amd_windows.json [hunk 3]
```diff
@@ -29,7 +29,11 @@
 	   "Unit_hipTextureMipmapObj2D_Check",
 	   "Unit_hipStreamPerThread_DeviceReset_1",
 	   "Unit_hipStreamPerThread_DeviceReset_2",
-	   "Unit_hipManagedKeyword_MultiGpu"	   
+	   "Unit_hipManagedKeyword_MultiGpu",
+	   "# Skipped due to EXSWCPHIPT-19",
```
Comment (FMarno) [inline tests/catch/hipTestMain/config/config_amd_windows.json]: convert to HIP_SKIP_TEST

File: tests/catch/unit/stream/hipStreamDestroy.cc [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+ *
+ * @param semaphore the semaphore to wait for.
+ */
+__global__ void waiting_kernel(int* semaphore) {
+  size_t tid = blockIdx.x * blockDim.x + threadIdx.x;
+  if (tid == 0) {
+    while (atomicCAS(semaphore, 1, 2) == 0) {
+    }
+  }
+}
```
Comment (FMarno) [inline tests/catch/unit/stream/hipStreamDestroy.cc]: find a different solution to this. No guarantee that both kernels will run concurrently so this could lead to deadlock.

**Review by FMarno — COMMENTED** [review]
Also need to rework Unit_hipStreamDestroy_WithPendingWork

**Comment by chrispaquot** [comment]
Any update on this?

**Comment by FMarno** [comment]
Hi @chrispaquot. @jatinx looking at a solution using managed memory but hit a bit of a snag with the atomics, more details are in ticket EXSWCPHIPT-78.

File: tests/catch/include/hip_test_common.hh [hunk 5]
```diff
@@ -329,11 +329,89 @@ struct MemTraits<MemcpyAsync> {
   }
 };
 
-}  // namespace HipTest
 
+namespace {
```
Comment (fabiomestre) [inline tests/catch/include/hip_test_common.hh]: The changes to this file will have to be removed since the same code will be commited in a separate PR

File: tests/catch/unit/stream/hipStreamDestroy.cc [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+  HIP_CHECK(hipFree(deviceData));
+}
+
+// hipStreamDestroy should return immediately then clean up the resources when the stream is empty
+// of work
+#if HT_AMD /* Disabled because frequency based wait is timing out on nvidia platforms */
+TEST_CASE("Unit_hipStreamDestroy_WithPendingWork") {
+#if HT_AMD
+  HipTest::HIP_SKIP_TEST(
+      "EXSWCPHIPT-44 - expected hipStreamDestroy to return immediately then release the resources "
```
Comment (fabiomestre) [inline tests/catch/unit/stream/hipStreamDestroy.cc:77]: Need further clarification on why EXSWCPHIPT-44 was rejected. And maybe update the skip test message accordingly

**Comment by FMarno** [comment]
@chrispaquot This is ready for review now. Unfortunately we've not been able to find a cross-platform reliable solution to replace the kernel that runs for a long time, but the current solution is at least reused between multiple tests. Due to that reuse, this PR is block by PR #2750.

**Comment by FMarno** [comment]
@chrispaquot @mangupta This is now unblocked and ready to merge
